enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -85,264 +85,367 @@ bool is_load_uncacheable(
-    const SmallPtrSetImpl<const Instruction *> &unnecessaryInstructions,
-    const std::map<Argument *, bool> &uncacheable_args, bool topLevel) {
-  bool mustcache = false;
-
-  // If the pointer operand is from an argument to the function, we need to
-  // check if the argument
-  //   received from the caller is uncacheable.
-  if (isa<UndefValue>(obj)) {
-  } else if (auto arg = dyn_cast<Argument>(obj)) {
-    auto found = uncacheable_args.find(arg);
-    if (found == uncacheable_args.end()) {
-      llvm::errs() << "uncacheable_args:\n";
-      for (auto &pair : uncacheable_args) {
-        llvm::errs() << " + " << *pair.first << ": " << pair.second
-                     << " of func " << pair.first->getParent()->getName()
-                     << "\n";
+struct CacheAnalysis {
+  AAResults &AA;
+  Function *oldFunc;
+  LoopInfo &OrigLI;
+  DominatorTree DT;
+  TargetLibraryInfo &TLI;
+  const SmallPtrSetImpl<const Instruction *> &unnecessaryInstructions;
+  const std::map<Argument *, bool> &uncacheable_args;
+  bool topLevel;
+  std::map<Value *, bool> seen;
+  CacheAnalysis(
+      AAResults &AA, Function *oldFunc, LoopInfo &OrigLI,
+      TargetLibraryInfo &TLI,
+      const SmallPtrSetImpl<const Instruction *> &unnecessaryInstructions,
+      const std::map<Argument *, bool> &uncacheable_args, bool topLevel)
+      : AA(AA), oldFunc(oldFunc), OrigLI(OrigLI), DT(*oldFunc), TLI(TLI),
+        unnecessaryInstructions(unnecessaryInstructions),
+        uncacheable_args(uncacheable_args), topLevel(topLevel) {}
+
+  bool is_value_mustcache_from_origin(Value *obj) {
+    if (seen.find(obj) != seen.end())
+      return seen[obj];
+
+    bool mustcache = false;
+
+    // If the pointer operand is from an argument to the function, we need to
+    // check if the argument
+    //   received from the caller is uncacheable.
+    if (isa<UndefValue>(obj) || isa<ConstantPointerNull>(obj)) {
+      return false;
+    } else if (auto arg = dyn_cast<Argument>(obj)) {
+      auto found = uncacheable_args.find(arg);
+      if (found == uncacheable_args.end()) {
+        llvm::errs() << "uncacheable_args:\n";
+        for (auto &pair : uncacheable_args) {
+          llvm::errs() << " + " << *pair.first << ": " << pair.second
+                       << " of func " << pair.first->getParent()->getName()
+                       << "\n";
+        }
+        llvm::errs() << "could not find " << *arg << " of func "
+                     << arg->getParent()->getName() << " in args_map\n";
       }
-      llvm::errs() << "could not find " << *arg << " of func "
-                   << arg->getParent()->getName() << " in args_map\n";
-    }
-    assert(found != uncacheable_args.end());
-    if (found->second) {
-      mustcache = true;
-    }
-  } else {
+      assert(found != uncacheable_args.end());
+      if (found->second) {
+        mustcache = true;
+        // EmitWarning("UncacheableOrigin", arg->getDebugLoc(), oldFunc,
```

Comment:
Is this warning still relevant?

Reply:
Potentially still relevant, though an Argument doesn't have debug info associated to attach to =/